Explore: Logs filtering in content outline

harisrozajac opened the pull request:
**What is this feature?**

This PR implements filtering in the content outline. 

**Background** 
Currently, the highlighting of content outline items (buttons) depends on the vertical scroll. When the vertical scroll is triggered, an effect is triggered that then looks at all content outline items and finds the first one whose position from the top is >= 0. We also have a logic that highlights the parent item if a child outline is selected and that section is collapsed. For example, the Queries section could have query A and query B as children, and if B is closest to the top, it is highlighted, but if the section is collapsed, then the parent will be highlighted.

### **Solution**

The issue with implementing filters is that the highlighting of the items would have to rely on the click state and not on the vertical scroll. After some deliberation with the UX and Product team, we've decided to get with the following option: 

**Filter and scroll into view**
When the user clicks on a filter, then logs should be filtered and the logs container pulled into view. This would also require highlighting by using a separate state because we can't highlight filter items based on scroll since they don't have a separate ref, and the logic that determines what should be highlighted when the scroll is triggered would always highlight the parent item. 

**Sync filter with logs volume panel**
The content outline is synced with the logs volume panel, so toggling filters in content outline will update the filters (series) in logs volume panel and vice versa. Highlighting multiple filters is possible with `cmd + click`

Note: When logs volume is collapsed filters in content outline are removed. This makes sense because we want to keep them in sync. 

**Demo**

https://github.com/grafana/grafana/assets/58232930/82dcb942-6aa5-4eb6-9f63-9fad50dc22d0



Closes: #86318 

**Special notes for your reviewer:**
- [ ] Works as expected from the user's perspective
- [ ] Test filtering in content outline with single clicks and cmd + click 
- [ ] Test filtering in logs volume panel and verify if the filters in content outline are updating
- [ ] Load a datasource that has no level labels (logs volume should show only one label) and verify that logs item in content outline doesn't have any children items

harisrozajac committed:
wip filtering

harisrozajac committed:
wip filtering

harisrozajac committed:
Merge branch 'main' into haris/explore/logs-in-content-outline

harisrozajac committed:
Remove corresponding children from parentlessItem ref on register

harisrozajac committed:
Remove corresponding children from parentlessItem ref on register

ifrost commented:
/deploy-to-hg

matyax commented:
We'll be looking at this shortly. We are currently at half capacity.

gelicia commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -152,6 +153,8 @@ class UnthemedLogs extends PureComponent<Props, State> {
   cancelFlippingTimer?: number;
   topLogsRef = createRef<HTMLDivElement>();
   logsVolumeEventBus: EventBus;
+  static contextType = ContentOutlineContext;
```
is this used?

harisrozajac replied:
yes, it's a way for class-based components to get access to Context without Context.Consumer component. otherwise it throws an error during render like this:
<img width="408" alt="Screenshot 2024-04-18 at 2 32 27 PM" src="https://github.com/grafana/grafana/assets/58232930/e3c1a85c-7cd7-4686-afd9-67a1cce8114e">

ifrost committed:
Highlighting filters

ifrost committed:
OMG

harisrozajac committed:
Merge branch 'main' into haris/explore/logs-in-content-outline

harisrozajac committed:
Handle updating highlight field inside context

ifrost committed:
Toggling legend from the content outline

harisrozajac committed:
Support multiple selects for log filters in content outline

harisrozajac committed:
Merge branch 'main' into haris/explore/logs-in-content-outline

harisrozajac committed:
Remove commented out code

harisrozajac commented:
/deploy-to-hg

harisrozajac committed:
Fix an issue with loki datasource; use same datastructure for logsVolume in Logs and LogVolume panel

harisrozajac commented:
@gelicia you were not doing anything wrong. this was a great catch! it turns out that logs volume panel does some additional transformations on the logs volume data available in logs panel. the data i had in logs volume panel worked fine with the test datasource, but it was different with gdev-loki. i fixed this and now it works as expected with loki

harisrozajac committed:
Fix registering logic

harisrozajac commented:
@gelicia there was a bug in the registering logic that I've fixed. it should be working now

harisrozajac committed:
cleanup

harisrozajac committed:
Don't register outline items when having multiple log volume levels

harisrozajac committed:
Merge branch 'main' into haris/explore/logs-in-content-outline

ifrost commented:
1) I used elastic, it returned one level and it was shown in the filters even though with just one filter it cannot be toggled. It didn't happen in loki though 🤔 

| loki | elastic |
|---|---|
| <img width="601" alt="Screenshot 2024-04-26 at 10 49 25" src="https://github.com/grafana/grafana/assets/745532/f7e74d5e-9fc7-483e-9c86-f5becfa07933"> | <img width="585" alt="Screenshot 2024-04-26 at 10 49 33" src="https://github.com/grafana/grafana/assets/745532/2796c82d-b129-4967-8690-8e043e73862f"> |

2) Not sure when exactly it happens, but sometimes clicking on the filter toggles the parent:

https://github.com/grafana/grafana/assets/745532/d6ad4dea-8778-4ad4-821e-7a1e0f4f6059

harisrozajac committed:
Remove checking mergeSingleChild logic because in context because ContentOutline does it

harisrozajac committed:
Fix logic so it works for elasticsearch

harisrozajac commented:
@ifrost:

> 1. I used elastic, it returned one level and it was shown in the filters even though with just one filter it cannot be toggled. It didn't happen in loki though

It turns out that `logsVolumeData.data` returns multiple items that when passed to `getLogLevelInfo()` they will return same labels. in this case there were 3 data frames and although they had different value.name values when run through `getLogLevelInfo` they return same label which was `unknown`. I'm not sure why this happens, but I've implemented a check inside `registerLogLevelsWithContentOutline` here https://github.com/grafana/grafana/pull/86431/commits/f01d49ae04bb3ab94304875baf314cfa4d7dd817, to check if multiple levels with same name are returned.

We do have a logic in ContentOutlineContext that checks if a duplicate sibling exists, but in this case, this doesn't help since this comparison runs only after the first child is already registered. I considered potentially doing something inside ContentOutline, like checking during rendering if there's only one child item and if that item is a filter, but I think it's better to not register one filter item at all. 

I've tested the fix with logs in testdata, loki, and elastic and everything is working as expected now.

> 2. Not sure when exactly it happens, but sometimes clicking on the filter toggles the parent:

I've fixed this here https://github.com/grafana/grafana/pull/86431/commits/f01d49ae04bb3ab94304875baf314cfa4d7dd817

matyax commented on public/app/features/explore/Logs/Logs.tsx:
```diff
@@ -203,6 +215,49 @@ class UnthemedLogs extends PureComponent<Props, State> {
     }
   }
 
+  registerLogLevelsWithContentOutline = () => {
```
I haven't checked out this locally, but did you check if this will be updated when we run a different query or change the. current query?

gelicia replied:
Confirmed if you have a query with `logfmt` that has the filters in the outline, then remove `logfmt` and re-run the query, the filters disappear

harisrozajac committed:
Better logic

ifrost commented:
We can improve some bits if we convert Logs.tsx into a functional component. This would allow us to get rid of ts-ignore

```
  // @ts-ignore
  private toggleLegendRef: React.MutableRefObject<(name: string, mode: SeriesVisibilityChangeMode) => void> =
    React.createRef();
```

--> we will use `useRef` hook which is a mutable ref.

Also we will be able to pull calculating log volume data out of LogsVolumePanelList (maybe to a context above Logs.tsx?) and just use the data provided instead of having additional manipulations like 

```
registerLogLevelsWithContentOutline = () => {
...
const numberOfLogVolumes = this.getNumberOfLogVolumes();
...
```

I've created a task for it in https://github.com/grafana/grafana/issues/87047. @grafana/observability-logs - is it okay if we work on it as the follow-up for this issue?

matyax commented:
Absolutely @ifrost ! Furthermore, I wouldn't mind contributing to the migration of this component from class to function. Just let me know.

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into haris/explore/logs-in-content-outline

gelicia reviewed (approved):
LGTM 🚀

ifrost reviewed (approved):
![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExenQ3MzB2cGkzdmc5djFlN2t6NzB2OHF5Z3EyNHFjN3IzcGdxNmUyNiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/huyVJYSKcArLiu8J5g/giphy.gif)

Elfo404 commented:
@grafana/observability-logs mind a sanity check on this one?

matyax commented:
On it.

matyax commented:
Checked this out locally and played with it a little bit. Some thoughts after that.

- Outline in general: this is something that has been on my mind for a few months already: I did not know/notice that the outline was there. Also "outline" doesn't mean anything to me (spanish native speaker with english technical knowledge), and I can't link with with anything more familiar. I discovered because @svennergr told me it was there.

Coming back to this PR, a video of me playing with it:

Uploading Demo.mov…

- The collapsed view is a bit tough for me. There's a lot of empty space, and I feel like it could be a bit smaller.

![Uploading imagen.png…]()

matyax commented:
Checked this out locally and played with it a little bit. Some thoughts after that.

- Outline in general: this is something that has been on my mind for a few months already: I did not know/notice that the outline was there. Also "outline" doesn't mean anything to me (spanish native speaker with english technical knowledge), and I can't link with with anything more familiar. I discovered because @svennergr told me it was there.

Coming back to this PR, a video of me playing with it:

https://github.com/grafana/grafana/assets/1069378/60bfdc84-0eb5-4e1a-8965-b16e098e425c


- The collapsed view is a bit tough for me. There's a lot of empty space, and I feel like it could be a bit smaller.

![Collapsed](https://github.com/grafana/grafana/assets/1069378/b66b4862-7434-49ff-a438-2759cd699703)

Like: 

![imagen](https://github.com/grafana/grafana/assets/1069378/a7386bab-53a3-48e9-924e-8007aade7d6c)

Even in the expanded mode, I find the empty space on the left a bit unpleasant.

![imagen](https://github.com/grafana/grafana/assets/1069378/9c3eaa65-806d-41b2-845d-d86d473ebf53)

- The log level filters are amazing! From the Logs perspective, I would love if:

  - They were color coded: red error, yellow warning, green info, grey unknown. Having other unexpected colors or no colors is very confusing.
    ![imagen](https://github.com/grafana/grafana/assets/1069378/905f75c0-df33-4497-8558-822c97f16bca)
  - In the collapsed way, I would love if they displayed a different icon other than "logs".
    ![imagen](https://github.com/grafana/grafana/assets/1069378/7cdf4ab6-400b-4491-8ea0-8d2130545038)
  - I like that we can filter by level, that's something we should have added a long time ago! What I'm not sure if that we should apply the filtering on the client-side or if we should modify the query when clicking in those actions, like we do with the 🔍  icons. For example, it can be very confusing that you were focused on warnings, but when you change the query that filter/selection is forgotten. E.g.  

https://github.com/grafana/grafana/assets/1069378/448f701a-2ae5-4fc0-92bc-36a34615d713

  - Lastly, if possible, it should be Error > Warning > Info > Unkown, in that order.
  
I don't see any reason not to merge this as it is, but it would be amazing if we can continue iterating on this to take the most value out of this feature for Explore users. Amazing work, thank you all!

matyax commented:
I also feel that the level filters should be togglable and combinable. The only way to return to the "everything selected" state is to click on a selected level. What if I want to see errors and warnings but not info?

Also, there seems to be a bug with scrolling and the selected level. Sometimes it only scrolls with the last option, and I lose the "logs" active state when clicking on some of the filters.

https://github.com/grafana/grafana/assets/1069378/669359cc-b7b6-4442-8843-448063c45451

gelicia commented:
For this PR, I will look into

- Adding the log color to the content outline
- Modifying the query to filter the logs by level
- Setting the order to Error, Warning, Info, Unknown.
- Multi-select filtering
- Bug with scrolling

For everything more broad, I will add to https://github.com/grafana/grafana/issues/87146 to be addressed later

For the above, if the implementation will be too heavy of a lift or will require iterating ideas with UX, I will ask the logs team first but may also put it into https://github.com/grafana/grafana/issues/87146 for later

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into haris/explore/logs-in-content-outline

gelicia commented:
> I also feel that the level filters should be togglable and combinable. The only way to return to the "everything selected" state is to click on a selected level. What if I want to see errors and warnings but not info?

This currently operates the same way as the logs volume legend filter. If we enhance one, we'd have to enhance the other. Is that something the logs team has discussed? Is it okay that we move this request to that and add to be sure the content outline filter works with that enhancement?

gelicia committed:
Add log color and sort

gelicia commented:
> Lastly, if possible, it should be Error > Warning > Info > Unknown, in that order.

This is not defined anywhere as far as I can tell. I ended up editing the LogsModel colors to put warning in this order, because it seemed as good of a place as any.

<img width="191" alt="Screenshot 2024-04-30 at 3 23 44 PM" src="https://github.com/grafana/grafana/assets/1533128/70e2fb72-2807-4b9f-8fad-5275899031aa">

diegoadams commented:
@matyax , thanks for the feedback, on this one:

> Even in the expanded mode, I find the empty space on the left a bit unpleasant.

We aligned with the megamenu to ensure consistency in the user experience across our menus, such as the placement of dropdowns, icons etc. 

 ![Uploading Screenshot 2024-05-01 at 17.12.05.png…]()

diegoadams commented:
> We aligned with the megamenu to ensure consistency in the user experience across our menus, such as the placement of dropdowns, icons etc.

diegoadams commented:
> This is not defined anywhere as far as I can tell. I ended up editing the LogsModel colors to put warning in this order, because it seemed as good of a place as any.

@gelicia @harisrozajac .  The colour code looks great! I recommend using the colours in their full saturation rather than faded, especially when they're in an active state.

<img width="1584" alt="Screenshot 2024-05-01 at 17 28 24" src="https://github.com/grafana/grafana/assets/95694434/342c7001-3083-44cd-b2ff-9f2b81e9b0fe">

diegoadams commented:
> * In the collapsed way, I would love if they displayed a different icon other than "logs".

@matyax , I agree, but the mini view has always presented a challenge due to its limited space. 

Take the main mega menu, for instance, where only Level 1 items display icons, and Level 2 items do not. In our design, we wanted to maintain a connection with the user, ensuring they are aware of their current context, whether it be logs, graphs, etc. 

For this phase, I would suggest we keep the log icons, and then in our next iteration, we can consider a different pattern.

diegoadams commented:
Sorry I clicked closed by mistake, I think its open again.

diegoadams commented:
@harisrozajac , could we keep these in a non-selected state when in their default? It appears cleaner and less cluttered. Also, given that only one filter can be applied at a time, it would make this functionality clearer to the user.

<img width="1492" alt="Screenshot 2024-05-01 at 19 03 36" src="https://github.com/grafana/grafana/assets/95694434/7ee401e3-ffff-4ca6-bf38-3b829b40bc02">

matyax commented:
@diegoadams Thanks for the clarifications! Personally, in every screenshot I see, the contents of the outline appear concentrated in a tiny upper-right section.

![Screenshot](https://github.com/grafana/grafana/assets/1069378/eeb4b6e8-2f78-4ebc-b25a-a68d6f21393f)

Maybe you can consider removing the padding on the left to clean up a bit of empty space?

![Option](https://github.com/grafana/grafana/assets/1069378/9c59264d-580c-4f00-a20b-2b907c5d10af)

Or move the upper level options to the left?

![Option 2](https://github.com/grafana/grafana/assets/1069378/b8e3906b-76b8-4917-9476-be2805db32c7)

In any case, I'm just sharing with you all my honest impressions on the Outline changes. Please feel free to ignore it.

---

Posting video of the scrolling issues I found again:


https://github.com/grafana/grafana/assets/1069378/eb023be5-7afc-4b42-9b0c-d0b2c8821749

---

Thank you all! Amazing work.

diegoadams commented:
> In any case, I'm just sharing with you all my honest impressions on the Outline changes. Please feel free to ignore it.

This is great feedback. Whether we make the changes or not, please always share your honest feedback. I appreciate it🙂

> Or move the upper-level options to the left?

We've limited the Content Outline to a maximum of 2 levels due to the limited horizontal space constraints. 
Therefore, placing Logs at level 2 would necessitate moving Log Filters to a new 3rd level.

Additionally, the dropdowns on the side are designed to potentially accommodate not just Logs but other data sources or new sections in the future.  


I'll get in touch with @amy-super, the UX lead for the Main Nav, to discuss if there is a slight adjustment we could make. But please don't wait on this as it may come in the next iteration.

diegoadams commented:
@gelicia @harisrozajac , I cant seem to access the https://ephemeral1511182186431harisro.grafana-dev.net/ anymore

gelicia commented:
/deploy-to-hg

gelicia commented:
> I'll get in touch with @amy-super, the UX lead for the Main Nav, to discuss if there is a slight adjustment we could make. But please don't wait on this as it may come in the next iteration.

Thanks for the discussion @diegoadams and @matyax - sounds like some good feedback for UX to iterate on. Feel free to add direction in #87146 since this PR will most likely be closed by then

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into haris/explore/logs-in-content-outline

gelicia committed:
Show items at full opacity

gelicia commented:
/deploy-to-hg

gelicia commented:
@diegoadams @matyax - removed the 30% opacity style so the colors should be nice and bright now. Let me know what you think!

gelicia committed:
Remove commented sort

diegoadams commented:
> @diegoadams - removed the 30% opacity style so the colors should be nice and bright now. Let me know what you think!

Thanks @gelicia🥳 , I can't seem to find gdev data source in the ephemeral to see what the logs filters look like. If you could send me a screenshot that would be great.

gelicia commented:
@diegoadams On the ephemeral, you can use this link https://ephemeral1511182186431harisro.grafana-dev.net/explore?schemaVersion=1&panes=%7B%22lpj%22:%7B%22datasource%22:%22grafanacloud-usage-insights%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Borg_id%3D%5C%22916%5C%22%7D%20%7C%3D%20%60%60%20%7C%20logfmt%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22grafanacloud-usage-insights%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1 

The datasource is grafanacloud-ephemeral1511182186431harisro-usage-insights and the query `{org_id="916"} |= `` | logfmt` shows some logs with different types.

I did just realize though that the non-logs content outline items lost their orange color, so I will need to fix that 😅

diegoadams commented:
@gelicia, haha no worries. 

I think we were going to have them not all show colours at the same time as only one filter can be applied at a time. Sorry I forgot to tag you in this comment:
https://github.com/grafana/grafana/pull/86431#issuecomment-2088185686

But having a look at it now, it works with the colours in full and also provides context within the mini view. So let's keep it like you have it! 

But having seen it now, I see that the full colours work well and gives the mini view clear context, so let's stick with your current version.

It looks fantastic Thanks, ! 🚀

gelicia commented:
I talked it over with @diegoadams and actually read his comment.. we will be changing this so by default, there will not be any colors showing next to the log levels. Only when one is clicked and the filter is active, will the active log level show its color. 

That and the scrolling bug are left to address.

matyax commented:
@gelicia The rearrangement of the colors looks great to me. That precedence should be followed in most places because it's also the precedence of error levels. Thanks for that!

Gave a quick test to this branch and it's looking great so far. Some extra thoughts, again, feel free to ignore.

Maybe capitalizing the levels and adding 2px more (theme.spacing(2.5)) to the left?

![imagen](https://github.com/grafana/grafana/assets/1069378/cc96d15e-d687-4cd5-a559-360be9f617d5)

matyax commented:
@gelicia The rearrangement of the colors looks great to me. That precedence should be followed in most places because it's also the precedence of error levels. Thanks for that!

Gave a quick test to this branch and it's looking great so far. Some extra thoughts, again, feel free to ignore.

Maybe capitalizing the levels and adding 2px more (theme.spacing(2.5)) to the left?

![imagen](https://github.com/grafana/grafana/assets/1069378/cc96d15e-d687-4cd5-a559-360be9f617d5)

// @grafana/observability-logs

gelicia committed:
Do not highlight if all are selected, and highlight = 100% opacity, active = 30%

gelicia commented:
/deploy-to-hg

gelicia committed:
dont auto collapse after filter unselect

gelicia committed:
Merge branch 'main' of https://github.com/grafana/grafana into haris/explore/logs-in-content-outline

gelicia commented:
I removed the logic that said it should auto-collapse if you remove the last filter - that didn't seem to make sense to me. Feel free to let me know that it was by design and I will add it back.

I'm not going to do the suggested changes to the spacing or with capitalization - the spacing being intentional to match other navigation components and the capitalization matching what is in the chart legend (just like clicking matches the behavior of the chart legend). I would rather leave these both uncapitalized but instead monospace the labels to align with the idea that the label is coming from the log line. This is obviously something to discuss with designers, and I do not want to make it part of this ticket. I have added both points of feedback to #87146

gelicia commented:
/deploy-to-hg

gelicia committed:
Remove console logs

gelicia commented:
/deploy-to-hg

gelicia committed:
No partial opacity, add height for consistent display

gelicia commented:
/deploy-to-hg

gelicia commented:
Talked with @diegoadams, who specified he wanted all active elements to have full opacity as well. That change was made and in the ephemeral. 

I'd like one more review from someone in @grafana/observability-logs - if there is feedback that was given and not addressed, it should be in #87146 . If you have feedback and it is not in there and you think it could be addressed later, please approve and add it to the other ticket. If you have feedback you would like implemented now or are not sure if it should be implemented now or later, please leave a comment and I will address it either by adding it to the other ticket or fixing it here.

matyax reviewed (approved):
Took a second look. The only feedback I have that's not mentioned in #87146 is that, without colors in the collapsed version, it's hard to know what each level icon means.

![imagen](https://github.com/grafana/grafana/assets/1069378/2386dfc4-350f-43b4-961d-6799f3050009)

gelicia commented:
> Took a second look. The only feedback I have that's not mentioned in #87146 is that, without colors in the collapsed version, it's hard to know what each level icon means.
> 
> ![imagen](https://private-user-images.githubusercontent.com/1069378/328488263-2386dfc4-350f-43b4-961d-6799f3050009.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwOTI5NTUsIm5iZiI6MTcxNTA5MjY1NSwicGF0aCI6Ii8xMDY5Mzc4LzMyODQ4ODI2My0yMzg2ZGZjNC0zNTBmLTQzYjQtOTYxZC02Nzk5ZjMwNTAwMDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDdUMTQzNzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzU0Zjc4MjhjMGYxODUyYTdlMTEyNzhjZWY3YjkzOWVhZjJiNTQ5MTU3ZjY0ODkwYjFmMTcwYzI1NzY1OWM4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.MX2VcOIeNfq58KmEdVsoS67iVkia0E9kIKmZ0l-Y1TQ)

This is directly implementing @diegoadams' [feedback](https://github.com/grafana/grafana/pull/86431#issuecomment-2088185686) "could we keep these in a non-selected state when in their default? It appears cleaner and less cluttered." Non selected state = no colors. The log level in collapsed are available via tooltip.

I am going to put this in the other ticket, as some iterations are necessary to find a good middle ground here.

matyax commented:
Brainstorming: how about displaying a single letter? E, W, I, D, U

gelicia commented:
> Brainstorming: how about displaying a single letter? E, W, I, D, U

Added to #87146
